artemis-server/src/main/java/org/apache/activemq/artemis/spi/core/security/scram/SCRAM.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.activemq.artemis.spi.core.security.scram;
+
+/**
+ * Defines sets of known SCRAM types with methods to fetch matching digest and hmac names
+ */
+public enum SCRAM {
+                   SHA1,
+                   SHA256,
+                   SHA512;
```

Comment:
If the SCRAM-SHA-512 support was removed it would seem this and various related bits below should go also.

(Personally I'd have left it and added some actual unit tests)

Reply:
Adding unit-test would require to implement the client-side as well.. that's a bit out-of-scope here for the moment. It could be added easily later though if desired.

Reply:
Adding unit tests for the code for the very functionality you are adding is clearly not 'out of scope'.

I've also already pointed you at some unit tests for this very same broker behaviour that you are adding, which could easily be used 'for inspiration'. 

You are also using a client here in your systest that has it, and which that the library you have 'taken inspiration from' actually lifted its own client impl from to begin with.

Reply:
For me adding a sasl-client implementation in the context that there is no AMQP-Client I'm aware of that support SHA-512 (yet) is at least too much hassle. The cupid client do not support SHA-512 and the "client" from the SASL-SCRAM-Lib only supports the authentication but not AMQP...

Reply:
Artemis includes a JMS client but I can't find how it is used, is there an example that uses AMQP+SASL+Artemis JMS Client?

Reply:
>>For me adding a sasl-client implementation in the context that there is no AMQP-Client I'm aware of that support SHA-512 (yet) is at least too much hassle.

My earlier comments about unit tests were about all the supported mechs, the point on the SHA-512 specifically was just that it would have then actually had some tests and could have remained (since the tests would after all be rather common, differing in the algorithm used and some expected values, i.e a likely trivial amount of additional effort). 

> Artemis includes a JMS client but I can't find how it is used, is there an example that uses AMQP+SASL+Artemis JMS Client?

There are vast numbers of examples (see the examples dir) and tests showing how the Artemis JMS client can be used, so I assume its something more nuanced you are asking? I'd note though that Artemis JMS client uses Artemis' own Core protocol, which doesnt support SASL.

Reply:
Yeah I specifically hoped there is an example specific to SASL usage (e.g. using SASL-Plain). That would be a good startingpoint as I then could simply use that example, change to SASL-SCRAM-SHA1 and debug where it fails.

Does 'Artemis JMS client uses Artemis own Core protocol' actually means that even not AMQP is used by the JMS client?

Reply:
The Artemis JMS client does not use AMQP, no. It uses the Artemis 'Core' protocol, which does not support SASL.

The Qpid JMS client uses AMQP 1.0, which does support SASL. It works with Artemis (and others) since it supports multiple protocols including AMQP 1.0.

Reply:
okay I think it makes sense then to delay SHA-512 support until theres actually a AMQP client supporting it.

Reply:
As cyrus-sasl seems to support it already that means there are actually AMQP 1.0 clients that support it (qpid-proton c/cpp/python/etc use cyrus-sasl), just not qpid-jms yet.